### DISCUSSION TITLE
How to get the residual vector after each nonlinear step

### DISCUSSION THREAD

**Discussion by tu1620808124** [firedrakeproject/firedrake#2891] [Firedrake support]
Hello,
In every SNES step， there will return a residual norm. When I do the debugging, I want to get the residual vector. PETSc maybe uses the post-check function. How to get the residual vector in firedrake？

**Comment by wence-** [comment]
You can pass in a `post_function_callback` function to the nonlinear solver that gets given the residual after it is assembled.

You can also set a custom monitor on the snes with `solver.snes.setMonitor` with signature `def monitor(snes, rnorm, its)` (IIRC) and then use `snes.getSolution` to get the current guess, and `snes.getFunction` to get the function for computing the residual (you can then call that function on the current guess to get the residual/
